Source/Core/InputCommon/ControllerInterface/WGInput/WGInput.cpp

```diff
@@ -0,0 +1,691 @@
+                  u32 haptics_index, std::string_view feedback_name)
+        : SimpleHaptics(haptics, feedback, haptics_index), m_feedback_name(feedback_name)
+    {
+    }
+    std::string GetName() const override
+    {
+      return fmt::format("{} {}", m_feedback_name, GetHapticsIndex());
+    }
+
+  private:
+    const std::string_view m_feedback_name;
+  };
+
+  void PopulateButtons()
+  {
+    // Using RawGameController for buttons because it gives us a nice array instead of a bitmask.
+    INT32 button_count = 0;
+    if (SUCCEEDED(m_raw_controller->get_ButtonCount(&button_count)))
+      m_buttons.resize(button_count);
+
+    u32 i = 0;
+    for (auto& button : m_buttons)
+    {
+      WGI::GameControllerButtonLabel lbl = WGI::GameControllerButtonLabel_None;
+      m_raw_controller->GetButtonLabel(i, &lbl);
+
+      if (lbl != WGI::GameControllerButtonLabel_None && lbl < wgi_button_names.size())
+        AddInput(new NamedButton(&button, wgi_button_names[lbl]));
+      else
+        AddInput(new IndexedButton(&button, i));
+
+      ++i;
+    }
+  }
+
+  void PopulateHaptics()
+  {
+    WGI::IRawGameController2* rgc2 = nullptr;
+    if (FAILED(m_raw_controller->QueryInterface(&rgc2)) || !rgc2)
+      return;
+
+    IVectorView<ABI::Windows::Devices::Haptics::SimpleHapticsController*>* haptics = nullptr;
+    if (FAILED(rgc2->get_SimpleHapticsControllers(&haptics)) || !haptics)
+      return;
+
+    unsigned int haptic_count = 0;
+    if (FAILED(haptics->get_Size(&haptic_count)))
+      return;
+
+    for (unsigned int h = 0; h != haptic_count; ++h)
+    {
+      ComPtr<ABI::Windows::Devices::Haptics::ISimpleHapticsController> haptic;
+      if (FAILED(haptics->GetAt(h, &haptic)))
+        continue;
+
+      IVectorView<ABI::Windows::Devices::Haptics::SimpleHapticsControllerFeedback*>* feedbacks =
+          nullptr;
+      if (FAILED(haptic->get_SupportedFeedback(&feedbacks)) || !feedbacks)
+        continue;
+
+      unsigned int feedback_count = 0;
+      if (FAILED(haptics->get_Size(&feedback_count)))
+        continue;
+
+      for (unsigned int f = 0; f != feedback_count; ++f)
+      {
+        ComPtr<ABI::Windows::Devices::Haptics::ISimpleHapticsControllerFeedback> feedback;
+        if (FAILED(feedbacks->GetAt(f, &feedback)))
+          continue;
+
+        UINT16 waveform = 0;
+        if (FAILED(feedback->get_Waveform(&waveform)))
+          continue;
+
+        std::string_view waveform_name{};
+
+        // Haptic Usage Page from HID spec.
+        switch (waveform)
+        {
+        case 0x1003:
+          waveform_name = "Click";
+          break;
+        case 0x1004:
+          waveform_name = "Buzz";
+          break;
+        case 0x1005:
+          waveform_name = "Rumble";
+          break;
+        }
+
+        if (!waveform_name.data())
+        {
+          WARN_LOG(PAD, "WGInput: Unknown haptics feedback waveform: %d.", waveform);
```

Comment:
All the logs should be converted to CONTROLLERINTERFACE.

Reply:
Done.